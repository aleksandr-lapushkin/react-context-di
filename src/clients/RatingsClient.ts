import { AxiosInstance } from "axios";

export interface RatingsClient {
    getRatingForUser(id: string): Promise<number>
}

export class AxiosRatingsClient implements RatingsClient {
    private axiosInstance: AxiosInstance;

    constructor(axiosInstance: AxiosInstance) {
        this.axiosInstance = axiosInstance
    }

    async getRatingForUser(id: string): Promise<number> {
        const result = await this.axiosInstance.get<{rating: number}>(`/api/users/${id}/rating`)

        if (result.data) {
            return result.data.rating
        }

        throw new Error("Unable to fetch Rating")
    }

}