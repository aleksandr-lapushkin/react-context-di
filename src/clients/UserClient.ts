import { AxiosInstance } from "axios";
import {ApiPaths} from "../constants/ApiPaths";

export interface User {
    id: string
    name: string
    fullName: string
}

export interface UserClient {
    getUser(id: string): Promise<User>
    getUsers(): Promise<User[]>
    getUserSafe(id: string): Promise<Result<User, ResultCode>>
}

export class AxiosUserClient implements UserClient {
    private axiosInstance: AxiosInstance;

    constructor(axiosInstance: AxiosInstance) {
        this.axiosInstance = axiosInstance;
    }

    async getUser(id: string): Promise<User> {
        try {
            const result = await this.axiosInstance.get<User>(ApiPaths.users.read(id))

            if (result.data) {
                return result.data;
            }
        } catch (err) {
            if (err?.response?.status === 404) {
                throw new Error("Not Found")
            }
            throw err
        }

        throw new Error("Failed to fetch User")
    }

    async getUsers(): Promise<User[]> {
        const result = await this.axiosInstance.get<{users: User[]}>(ApiPaths.users.list)

        if (result.data) {
            return result.data.users;
        }

        throw new Error("Failed to fetch Users")
    }

    async getUserSafe(id: string): Promise<Result<User, ResultCode>> {
        try {
            const result = await this.axiosInstance.get<User>(ApiPaths.users.read(id))
            return new SuccessImpl<User>(result.data)

        } catch (err) {
            if (err?.response?.status === 404) {
                return new ServerFailureImpl<ResultCode>(ResultCode.NOT_FOUND)
            }
            if (err?.response?.data) {
                return new ServerFailureImpl(ResultCode.UNKNOWN)
            }
            return new RequestFailureImpl(err.toString())
        }
    }

}












//Bonus stuff
export enum ResultCode {
    OK,
    NOT_FOUND,
    UNKNOWN
}
export enum ResultType {
    SUCCESS,
    SERVER_FAILURE,
    REQUEST_FAILURE
}
interface Success<T> extends Fold<T, any>{
    data: T,
    type: ResultType.SUCCESS
}
interface SuccessFolder<T, O> {
    (data: T): O
}
interface ServerFailure<E> extends Fold<any, E>{
    error: E,
    type: ResultType.SERVER_FAILURE
}
interface ServerFailureFolder<E, O> {
    (data: E): O
}
interface RequestFailure extends Fold<any, any> {
    message?: string,
    type: ResultType.REQUEST_FAILURE
}
interface RequestFailureFolder<O> {
    (data?: string): O
}
interface Fold<T, E> {
    fold: <O>(s: SuccessFolder<T, O>, sf: ServerFailureFolder<E, O>, rf: RequestFailureFolder<O>) => O
}
export type Result<T, E> = (Success<T> | ServerFailure<E> | RequestFailure) & Fold<T, E>
export class SuccessImpl<T> implements Success<T> {
    data: T;
    type: ResultType.SUCCESS;

    constructor(data: T) {
        this.type = ResultType.SUCCESS
        this.data = data
    }

    public fold<O>(s: SuccessFolder<T, O>, sf: ServerFailureFolder<any, O>, rf: RequestFailureFolder<O>): O {
        return s(this.data)
    }
}
export class ServerFailureImpl<E> implements ServerFailure<E> {
    error: E;
    type: ResultType.SERVER_FAILURE;
    constructor(error: E) {
        this.type = ResultType.SERVER_FAILURE;
        this.error = error;
    }
    fold<O>(s: SuccessFolder<any, O>, sf: ServerFailureFolder<E, O>, rf: RequestFailureFolder<O>): O {
        return sf(this.error)
    }
}
export class RequestFailureImpl implements RequestFailure {
    message?: string;
    type: ResultType.REQUEST_FAILURE;
    constructor(message?: string) {
        this.message = message;
        this.type = ResultType.REQUEST_FAILURE;
    }
    fold<O>(s: SuccessFolder<any, O>, sf: ServerFailureFolder<any, O>, rf: RequestFailureFolder<O>): O {
        return rf(this.message)
    }

}