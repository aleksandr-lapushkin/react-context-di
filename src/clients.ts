import {AxiosUserClient, UserClient} from "./clients/UserClient";
import axios from "axios";
import {AxiosRatingsClient, RatingsClient} from "./clients/RatingsClient";

let axiosInstance = axios.create();
export const usersClient: UserClient = new AxiosUserClient(axiosInstance)
export const ratingsClient: RatingsClient = new AxiosRatingsClient(axiosInstance)