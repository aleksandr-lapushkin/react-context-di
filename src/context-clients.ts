import {AxiosUserClient, UserClient} from "./clients/UserClient";
import {AxiosRatingsClient, RatingsClient} from "./clients/RatingsClient";
import axios from "axios";
import React, {useContext} from "react";

interface Clients {
    usersClient: UserClient
    ratingsClient: RatingsClient
}

let axiosInstance = axios.create();
export const ClientsContext = React.createContext<Clients>({
    usersClient: new AxiosUserClient(axiosInstance),
    ratingsClient: new AxiosRatingsClient(axiosInstance)
})

export function useClient<K extends keyof Clients>(key: K): Clients[K] {
    const context = useContext(ClientsContext)

    return context[key]
}