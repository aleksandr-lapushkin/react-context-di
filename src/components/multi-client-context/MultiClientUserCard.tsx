import {UserCard} from "../display/UserCard";
import {useClient} from "../../context-clients";
import {User} from "../../clients/UserClient";
import React, {useEffect, useState} from "react";
import {UserRating} from "../display/UserRating";
import {UserAvatar} from "../display/UserAvatar";
import {MultiClientRating} from "./MultiClientRating";

export interface MultiClientUserCardProps {
    userId: string
}

export const MultiClientUserCard: React.FC<MultiClientUserCardProps> = ({userId}) => {
    const userClient = useClient("usersClient");
    const [user, setUser] = useState<User>();
    const [error, setError] = useState<string>();

    useEffect(() => {
        const load = async () => {
            try {
                const data = await userClient.getUser(userId)
                setUser(data)
            } catch (err) {
                setError(err.message)
            }
        }
        load()
    })

    return <UserCard
        user={user}
        titleRenderer={u => <div><MultiClientRating userId={u.id}><UserAvatar/></MultiClientRating><span style={{margin: '16px'}}>{u.name}</span></div>}
    />
}