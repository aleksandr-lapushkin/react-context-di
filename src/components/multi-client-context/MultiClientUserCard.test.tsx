import {StaticRouter} from "react-router";
import { render } from "@testing-library/react";
import React, {PropsWithChildren} from "react";
import {Clients, ClientsContext} from "../../context-clients";
import {UserClient} from "../../clients/UserClient";
import {mock, instance, when} from "ts-mockito";
import {RatingsClient} from "../../clients/RatingsClient";
import {MultiClientUserCard} from "./MultiClientUserCard";

describe("Multi-client Context-backed UserCard", () => {
    const usersClient: UserClient = mock<UserClient>()
    const ratingsClient: RatingsClient = mock<RatingsClient>()

    it("Can show single User's data", async () => {
        when(usersClient.getUser("1")).thenResolve({id: "1", name: "test-user", fullName: "Test User"})
        when(ratingsClient.getRatingForUser("1")).thenResolve(42)

        const element = render(
            <ClientsContext.Provider value={{
                usersClient: instance(usersClient),
                ratingsClient: instance(ratingsClient)}}
            >
                <MultiClientUserCard userId={"1"}/>
            </ClientsContext.Provider>
        )

        const ratingElement = await element.findByTitle("42")
        expect(ratingElement).toBeInTheDocument()
        const nameElement = await element.findByText("test-user")
        expect(nameElement).toBeInTheDocument()

        expect(element.container).toMatchSnapshot()
    })

})