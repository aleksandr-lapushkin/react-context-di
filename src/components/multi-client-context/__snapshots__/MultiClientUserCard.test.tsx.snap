// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Multi-client Context-backed UserCard Can show single User's data 1`] = `
<div>
  <div
    class="ant-card ant-card-bordered"
  >
    <div
      class="ant-card-head"
    >
      <div
        class="ant-card-head-wrapper"
      >
        <div
          class="ant-card-head-title"
        >
          <div>
            <div>
              <span
                class="ant-badge"
              >
                <span
                  class="ant-avatar ant-avatar-lg ant-avatar-circle ant-avatar-image"
                >
                  <img
                    src="https://via.placeholder.com/150"
                  />
                </span>
                <sup
                  class="ant-scroll-number ant-badge-count ant-badge-multiple-words"
                  data-show="true"
                  title="42"
                >
                  <span
                    class="ant-scroll-number-only"
                    style="transform: translateY(-400%);"
                  >
                    <p
                      class=""
                    >
                      0
                    </p>
                    <p
                      class=""
                    >
                      1
                    </p>
                    <p
                      class=""
                    >
                      2
                    </p>
                    <p
                      class=""
                    >
                      3
                    </p>
                    <p
                      class="current"
                    >
                      4
                    </p>
                    <p
                      class=""
                    >
                      5
                    </p>
                    <p
                      class=""
                    >
                      6
                    </p>
                    <p
                      class=""
                    >
                      7
                    </p>
                    <p
                      class=""
                    >
                      8
                    </p>
                    <p
                      class=""
                    >
                      9
                    </p>
                    <p
                      class=""
                    >
                      0
                    </p>
                    <p
                      class=""
                    >
                      1
                    </p>
                    <p
                      class=""
                    >
                      2
                    </p>
                    <p
                      class=""
                    >
                      3
                    </p>
                    <p
                      class=""
                    >
                      4
                    </p>
                    <p
                      class=""
                    >
                      5
                    </p>
                    <p
                      class=""
                    >
                      6
                    </p>
                    <p
                      class=""
                    >
                      7
                    </p>
                    <p
                      class=""
                    >
                      8
                    </p>
                    <p
                      class=""
                    >
                      9
                    </p>
                    <p
                      class=""
                    >
                      0
                    </p>
                    <p
                      class=""
                    >
                      1
                    </p>
                    <p
                      class=""
                    >
                      2
                    </p>
                    <p
                      class=""
                    >
                      3
                    </p>
                    <p
                      class=""
                    >
                      4
                    </p>
                    <p
                      class=""
                    >
                      5
                    </p>
                    <p
                      class=""
                    >
                      6
                    </p>
                    <p
                      class=""
                    >
                      7
                    </p>
                    <p
                      class=""
                    >
                      8
                    </p>
                    <p
                      class=""
                    >
                      9
                    </p>
                  </span>
                  <span
                    class="ant-scroll-number-only"
                    style="transform: translateY(-200%);"
                  >
                    <p
                      class=""
                    >
                      0
                    </p>
                    <p
                      class=""
                    >
                      1
                    </p>
                    <p
                      class="current"
                    >
                      2
                    </p>
                    <p
                      class=""
                    >
                      3
                    </p>
                    <p
                      class=""
                    >
                      4
                    </p>
                    <p
                      class=""
                    >
                      5
                    </p>
                    <p
                      class=""
                    >
                      6
                    </p>
                    <p
                      class=""
                    >
                      7
                    </p>
                    <p
                      class=""
                    >
                      8
                    </p>
                    <p
                      class=""
                    >
                      9
                    </p>
                    <p
                      class=""
                    >
                      0
                    </p>
                    <p
                      class=""
                    >
                      1
                    </p>
                    <p
                      class=""
                    >
                      2
                    </p>
                    <p
                      class=""
                    >
                      3
                    </p>
                    <p
                      class=""
                    >
                      4
                    </p>
                    <p
                      class=""
                    >
                      5
                    </p>
                    <p
                      class=""
                    >
                      6
                    </p>
                    <p
                      class=""
                    >
                      7
                    </p>
                    <p
                      class=""
                    >
                      8
                    </p>
                    <p
                      class=""
                    >
                      9
                    </p>
                    <p
                      class=""
                    >
                      0
                    </p>
                    <p
                      class=""
                    >
                      1
                    </p>
                    <p
                      class=""
                    >
                      2
                    </p>
                    <p
                      class=""
                    >
                      3
                    </p>
                    <p
                      class=""
                    >
                      4
                    </p>
                    <p
                      class=""
                    >
                      5
                    </p>
                    <p
                      class=""
                    >
                      6
                    </p>
                    <p
                      class=""
                    >
                      7
                    </p>
                    <p
                      class=""
                    >
                      8
                    </p>
                    <p
                      class=""
                    >
                      9
                    </p>
                  </span>
                </sup>
              </span>
              <span
                style="margin: 16px;"
              >
                test-user
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="ant-card-body"
    >
      <div>
        <div>
          ID: 
          1
        </div>
        <div>
          Full Name: 
          Test User
        </div>
      </div>
    </div>
  </div>
</div>
`;
